### PR TITLE
Split up BPMNElementContainerProcessor.onElementCompleted into two methods

### PR DESCRIPTION
**Pull request by pihme**
## Description

The problem with the code was that this method was called after the element had already completed.
Therefore, the element and all its variables were already removed from storage.
The change introduces two methods before/after completion of the element. This allows to separate steps
that need to happen before the element is deleted from the steps that can happen after the element is deleted.

The method was also renamed from onChildCompleted which sournds like it should be called whenever an
element was completed - which was not the case - to before/AfterExecutionPathCompleting which better
reflects the semantics of the method.

## Related issues

related to #6172 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [X] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [X] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: engine/src/main/java/io/zeebe/engine/processing/bpmn/BpmnElementContainerProcessor.java [hunk 1]
```diff
@@ -35,14 +35,27 @@ void onChildActivating(
...
+   * The execution path of a child element is about to be completed.
    *
    * @param element the instance of the BPMN element container
    * @param flowScopeContext process instance-related data of the element container
-   * @param childContext process instance-related data of the child element that is completed
+   * @param childContext process instance-related data of the child element that is completed. At
+   *     this point in time the element is still present in the state
    */
-  void onChildCompleted(
+  void beforeExecutionPathCompleting(
```
Comment (korthout) [inline engine/src/main/java/io/zeebe/engine/processing/bpmn/BpmnElementContainerProcessor.java:45]: I understand the split into before and after completing, however we already have concepts for this: Completing and Completed. Completing is the state of the element when it starts to complete, while completed is the state of the element when it has completed. I think we can use `onExecutionPathCompleting` and `onExecutionPathCompleted` respectively.

File: engine/src/main/java/io/zeebe/engine/processing/bpmn/behavior/BpmnStateTransitionBehavior.java [hunk 2]
```diff
@@ -164,6 +164,23 @@ private void verifyIncidentResolving(final BpmnElementContext context, final Str
     }
   }
 
+  public <T extends ExecutableFlowNode>
+      BpmnElementContext transitionToCompletedWithParentNotification(
```
Comment (korthout) [inline engine/src/main/java/io/zeebe/engine/processing/bpmn/behavior/BpmnStateTransitionBehavior.java:168]: I'm wondering whether this method should just be merged into transitionToCompleted, but let's keep it like this for now.

File: engine/src/main/java/io/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java [hunk 3]
```diff
@@ -199,50 +202,27 @@ public void onChildActivating(
...
-        .getOutputElement()
-        .flatMap(Expression::getVariableName)
-        .map(BufferUtil::wrapString)
-        .ifPresent(
-            variableName -> stateBehavior.setLocalVariable(childContext, variableName, NIL_VALUE));
-
-    stateBehavior.setLocalVariable(
-        childContext, LOOP_COUNTER_VARIABLE, wrapLoopCounter(loopCounter));
-  }
-
```
Comment (korthout) [inline engine/src/main/java/io/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java:229]: Normally this is hard to review, but we already talked about it (automated method reordering). Up to you whether you still want to move it into another commit

File: engine/src/main/java/io/zeebe/engine/processing/bpmn/container/CallActivityProcessor.java [hunk 4]
```diff
@@ -132,7 +137,7 @@ public void onChildActivating(
       final BpmnElementContext childContext) {}
 
   @Override
-  public void onChildCompleted(
+  public void beforeExecutionPathCompleting(
```
Comment (korthout) [inline engine/src/main/java/io/zeebe/engine/processing/bpmn/container/CallActivityProcessor.java:140]: In ProcessProcessor, you've moved the current implementation of onChildCompleted into the after method. Is it necessary here to move it into the before? Otherwise, I'd prefer consistency and start by having this code in the method that matches most with the completed state (i.e. after)

File: engine/src/test/java/io/zeebe/engine/processing/bpmn/gateway/ExclusiveGatewayTest.java [hunk 5]
```diff
@@ -308,14 +308,14 @@ public void shouldEndScopeIfGatewayHasNoOutgoingFlows() {
...
-            ProcessInstanceIntent.ELEMENT_COMPLETING,
-            ProcessInstanceIntent.ELEMENT_COMPLETED,
-            ProcessInstanceIntent.ELEMENT_COMPLETING,
-            ProcessInstanceIntent.ELEMENT_COMPLETED);
+            tuple(BpmnElementType.EXCLUSIVE_GATEWAY, ProcessInstanceIntent.ELEMENT_ACTIVATING),
+            tuple(BpmnElementType.EXCLUSIVE_GATEWAY, ProcessInstanceIntent.ELEMENT_ACTIVATED),
+            tuple(BpmnElementType.EXCLUSIVE_GATEWAY, ProcessInstanceIntent.ELEMENT_COMPLETING),
+            tuple(BpmnElementType.EXCLUSIVE_GATEWAY, ProcessInstanceIntent.ELEMENT_COMPLETED),
+            tuple(BpmnElementType.PROCESS, ProcessInstanceIntent.ELEMENT_COMPLETING),
+            tuple(BpmnElementType.PROCESS, ProcessInstanceIntent.ELEMENT_COMPLETED));
```
Comment (korthout) [inline engine/src/test/java/io/zeebe/engine/processing/bpmn/gateway/ExclusiveGatewayTest.java:318]: This change can nicely be moved into a separate commit describing why it was useful.

**Review by korthout — APPROVED** [review]
Looks good in general, so here's already my approval 🎉  I do have a couple of suggestions though.

First, I think the PR description can be improved. In it you mention:
>The problem with the code was that this method was called after the element had already completed. Therefore, the element and all its variables were already removed from storage.

But that only becomes true as soon as the specific element processor is migrated. In the current situation, the element is removed on processing of the Element_Completed event. The rest of the description is clear in my opinion.

Then I have some other suggestions about naming. If you do happen to take these along, then please update the description and commit messages accordingly.

Lastly, I suggest you shorten the commit title to something like: `refactor(engine): split onElementCompleted in 2 methods`.

File: engine/src/test/java/io/zeebe/engine/processing/bpmn/gateway/ExclusiveGatewayTest.java (hunk 5 above)
Comment (pihme) [reply engine/src/test/java/io/zeebe/engine/processing/bpmn/gateway/ExclusiveGatewayTest.java:318]: won't do. I think this is self evident and requires no explanation

File: engine/src/main/java/io/zeebe/engine/processing/bpmn/container/CallActivityProcessor.java (hunk 4 above)
Comment (pihme) [reply engine/src/main/java/io/zeebe/engine/processing/bpmn/container/CallActivityProcessor.java:140]: Yes, it is because of `stateBehavior.propagateTemporaryVariables(childContext, callActivityContext);`

File: engine/src/main/java/io/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java (hunk 3 above)
Comment (pihme) [reply engine/src/main/java/io/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java:229]: won't do

File: engine/src/main/java/io/zeebe/engine/processing/bpmn/BpmnElementContainerProcessor.java (hunk 1 above)
Comment (pihme) [reply engine/src/main/java/io/zeebe/engine/processing/bpmn/BpmnElementContainerProcessor.java:45]: Naming discussed f2f
